src/Builder.js

```diff
@@ -24,44 +24,39 @@ class Builder {
 
     const groups = this._buildPolygonGroups(navMesh);
 
-    zone.groups = [];
-
-    const findPolygonIndex = function (group, p) {
-      for (let i = 0; i < group.length; i++) {
-        if (p === group[i]) return i;
-      }
-    };
-
-    // TODO: This block represents 50-60% of navigation mesh construction time,
+    // TODO: This block represents a large portion of navigation mesh construction time
     // and could probably be optimized. For example, construct portals while
     // determining the neighbor graph.
-    groups.forEach((group) => {
+    zone.groups = new Array(groups.length);
+    groups.forEach((group, groupIndex) => {
 
-      const newGroup = [];
+      const indexByPolygonId = {};
+      group.forEach((poly, polyIndex) => { indexByPolygonId[poly.id] = polyIndex; });
 
-      group.forEach((p) => {
+      const newGroup = new Array(group.length);
```

Comment:
I can't measure any difference pre-allocating these arrays, and (excluding typed arrays) it's rarely done in JS in my experience. Is there a reason not to just do `const newGroup = []` here?

Reply:
No, I just did it out of habit.